### PR TITLE
Update Autorest C#

### PR DESCRIPTION
**Pull request by pakrym**
Major changes
1. Required properties for input-output models are settable to allow for Read-Modify-Write scenarios.
1. Rest clients are renamed along with main ones

### CODE REVIEW

File: sdk/search/Azure.Search.Documents/src/Generated/Models/AnalyzeRequest.cs [hunk 1]
```diff
@@ -41,7 +41,7 @@ internal AnalyzeRequest(string text, AnalyzerName? analyzer, TokenizerName? toke
         }
 
         /// <summary> The text to break into tokens. </summary>
-        public string Text { get; }
+        public string Text { get; set; }
```
Comment (heaths) [inline sdk/search/Azure.Search.Documents/src/Generated/Models/AnalyzeRequest.cs:44]: From @KrzysztofCwalina, required properties should only be settable from within the constructor.

Comment (heaths) [reply sdk/search/Azure.Search.Documents/src/Generated/Models/AnalyzeRequest.cs:44]: Same applies below.

**Comment by heaths** [comment]
This is a departure from previous conversations with @KrzysztofCwalina. I see the point, but then how do we validate required parameters are provided, especially if they end up being *required path* parameters? Some are, and if not provided could cause issues (think back to the problem I demonstrated with path separator canonicalization and getting different response models with Key Vault).

Also, is this going to respect your new attributes for collection properties? I see setters were added to some models that only had a get accessor, though I hadn't customized yet. When I do, will this respect not wanting a setter (per guidelines)?

**Comment by pakrym** [comment]
> This is a departure from previous conversations with @KrzysztofCwalina. I see the point, but then how do we validate required parameters are provided, especially if they end up being required path parameters? Some are, and if not provided could cause issues (think back to the problem I demonstrated with path separator canonicalization and getting different response models with Key Vault).

Maybe we would have to generate setters that validate the value being not null the same way we do in constructors. And models can easily be customized to remove the public setters where appropriate.

The current pattern causes huge problems for largish models with a few required properties that could be update:

``` C#
var properties = xxxClient.Get();
var updatedProperties = new VaultProperties(properties.TenantId, Sku.Standard)
{
AccessPolicies = properties.accessPolicies;
VaultUri = properties.vaultUri;
EnabledForDeployment = properties.enabledForDeployment;
EnabledForDiskEncryption = properties.enabledForDiskEncryption;
EnabledForTemplateDeployment = properties.enabledForTemplateDeployment;
EnableSoftDelete = properties.enableSoftDelete;
SoftDeleteRetentionInDays = properties.softDeleteRetentionInDays;
EnableRbacAuthorization = properties.enableRbacAuthorization;
CreateMode = properties.createMode;
EnablePurgeProtection = properties.enablePurgeProtection;
NetworkAcls = properties.networkAcls;
PrivateEndpointConnections = properties.privateEndpointConnections;

}
xxxClient.Update(updatedProperties);
```

I'll make sure to follow up with @KrzysztofCwalina 

>Also, is this going to respect your new attributes for collection properties? I see setters were added to some models that only had a get accessor, though I hadn't customized yet. When I do, will this respect not wanting a setter (per guidelines)?

Just the fact that you would redefine a property without a public setter would be enough for generated code not to have public setter defined. The initialization attributes would work fine as well.

**Comment by pakrym** [comment]
Chatted with  @KrzysztofCwalina he's fine with the change.

**Comment by heaths** [comment]
Okay. I agree with the concept, certainly. Can we (perhaps follow-up PRs) get valudating properties then? I like that idea. Otherwise, looks good.

**Review by heaths — APPROVED** [review]

**Comment by pakrym** [comment]
Filed https://github.com/Azure/autorest.csharp/issues/707 to follow up.

**Pull request merged**